tests/unit/lms/services/group_info_test.py

```diff
@@ -0,0 +1,124 @@
+from unittest import mock
+
+import pytest
+
+from lms.models import ApplicationInstance, GroupInfo
+from lms.services.group_info import GroupInfoService
+
+
+class TestGroupInfoUpsert:
+    AUTHORITY = "TEST_AUTHORITY_PROVIDED_ID"
+    CONSUMER_KEY = "TEST_CONSUMER_KEY"
+
+    def test_it_adds_a_new_GroupInfo_if_none_exists(
+        self, db_session, group_info_svc, params
+    ):
+        group_info_svc.upsert(
+            authority_provided_id=self.AUTHORITY,
+            consumer_key=self.CONSUMER_KEY,
+            params=params,
+        )
+
+        group_info = self._get_inserted_group_info(db_session)
+
+        assert group_info.consumer_key == self.CONSUMER_KEY
+        assert group_info.context_title == params["context_title"]
+        assert group_info.context_label == params["context_label"]
+
+    def test_it_updates_an_existing_GroupInfo_if_one_already_exists(
+        self, db_session, group_info_svc, params
+    ):
+        db_session.add(
+            GroupInfo(
+                **dict(
+                    params,
+                    id=None,
+                    authority_provided_id=self.AUTHORITY,
+                    consumer_key=self.CONSUMER_KEY,
+                )
+            )
+        )
+
+        group_info_svc.upsert(
+            authority_provided_id=self.AUTHORITY,
+            consumer_key=self.CONSUMER_KEY,
+            params=dict(params, context_title="NEW_TITLE"),
+        )
+
+        group_info = self._get_inserted_group_info(db_session)
+
+        assert group_info.consumer_key == self.CONSUMER_KEY
+        assert group_info.context_label == params["context_label"]
+        assert group_info.context_title == "NEW_TITLE"
+
+    def test_it_ignores_non_metadata_params(self, db_session, group_info_svc, params):
```

Comment:
```suggestion
    def test_it_ignores_authority_provided_id(self, db_session, group_info_svc, params):
```

This doesn't actually test that `id` is ignored

Reply:
I wonder if this is the correct way to fix this. The original name of the function states the intention of the test. If that isn't implemented well, then the test should probably change.

Reply:
I don't think I necessarily phrased that very well. What I mostly mean here, is that the name of a test function indicates what it is trying to prove, not necessarily how it does it.

For example `test_it_makes_it_upper_case()` will test some specific examples of upper case, but not all of them. So changing it to `test_it_makes_these_specific_strings_upper_case()` is more accurate description of the _implementation_, but a less accurate description of the _intention_. It also becomes very hard to think of how to write that test better if it just describes itself (it's kind of circular).

In most cases it's not really possible to test all possible inputs, so most tests necessarily pick some special values or circumstances to stand in for the general case. That's not actually true here as it's trivial to cover both options, so I'll add that in.

Reply:
Yes, that makes sense. Also realising this test was a perfect `@pytest.mark.parametrize` opportunity